### ISSUE TITLE
To do lists

### ISSUE THREAD

**Issue by OscarGodson** [chjj/marked#107]
https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments

To be honest, I'm not entirely sure if Markdown should support this or not. Seems very, very Github centric, but curious what your thoughts are.

**Comment by toolbear** [comment]
:+1: 

For another perspective, I'm coming at this as a user of marked's lexer. My use case is extracting complete and incomplete tasks from `README.md`, `TODO.md`, and GitHub pull request descriptions.

I'm hoping task list support isn't implemented exclusively in the renderer. Something akin to this from the lexer:

```
markdown        | `completed` property of | `typeof token.completed` | `task` property of
                | list_item_start token   |                          | corresponding text token
----------------|-------------------------|--------------------------|------------------------
`* an item`     |                         | 'undefined'              | an item
`* [x] did it   | true                    | 'boolean'                | did it
`* [v] this too | true                    | 'boolean'                | this too
`* [ ] to-do    | false                   | 'boolean'                | to-do
```

This would be a non-breaking change. Custom renderers can opt-in by testing for a defined value of the `completed` property. We avoid a proliferation of "strip off the ballot box markdown" code by providing the already stripped text in the `task` property.

## Example

```coffee
doc = """
* [x] my completed task
* [ ] my incomplete task
"""

coffee> new Lexer({ gfm: false }).lex(doc)
[ { type: 'list_start', ordered: false },
  { type: 'list_item_start' },
  { type: 'text',
    text: '[x] my completed task' },
  { type: 'list_item_end' },
  { type: 'list_item_start' },
  { type: 'text',
    text: '[ ] my incomplete task' },
  { type: 'list_item_end' },
  { type: 'list_end' },
  links: {} ]

coffee> new Lexer({ gfm: true }).lex(doc)
[ { type: 'list_start', ordered: false },
  { type: 'list_item_start', completed: true },
  { type: 'text',
    text: '[x] my completed task',
    task: 'my completed task' },
  { type: 'list_item_end' },
  { type: 'list_item_start', completed: false },
  { type: 'text',
    text: '[ ] my incomplete task',
    task: 'my incomplete task' },
  { type: 'list_item_end' },
  { type: 'list_end' },
  links: {} ]
```

---

This is a pretty big request on my part. If you're favorable to the idea of changing the lexer I'd be happy to work on a PR for just that part, leaving rendering alone for now.

**Comment by anaran** [comment]
github task list support is not implemented yet, right?

It's kind of hard to follow the situation about them.

This issue calling them `To do lists` does not help either with explorability of the subject matter.

Rename it to `github task lists` please.

**Comment by dwendt** [comment]
It's a GFM task list if you want to be pedantic. Simply "GitHub task list" implies some integration into github's systems.

**Comment by toolbear** [comment]
If you really want to be pedantic, it *is* integrated into GitHub's systems: it has special rendering in all places Markdown is accepted; in a few places the task list is interactive.

**Comment by jhollingworth** [comment]
I've just opened #587 which should resolve this issue

**Comment by Reificator** [comment]
👍 for unicode substitution.  I use a great many things that take advantage of marked, including Visual Studio Code's markdown preview.  Would love the ability to render a ☐ and a ☑ and maybe even a ☒.

https://en.wikipedia.org/wiki/Checkbox#Unicode

**Comment by jungor** [comment]
This works for me...
```jsx
renderer.listitem = function (text, level) {
  let isCheckedTaskItem = checkedTaskItemPtn.test(text);
  let isUncheckedTaskItem = uncheckedTaskItemPtn.test(text);
  if (isCheckedTaskItem) text = text.replace(checkedTaskItemPtn, '<i class="fa fa-check-square" aria-hidden="true"></i>')+'\n';
  if (isUncheckedTaskItem) text = text.replace(uncheckedTaskItemPtn, '<i class="fa fa-square-o" aria-hidden="true"></i>')+'\n';
  let cls = (isCheckedTaskItem || isUncheckedTaskItem) ? ' class="todo-list-item"' : '';
  return '<li'+ cls + '>' + text + '</li>\n';
};
```
I use icons from `fontawesome`, but can just use html checkbox simply, of course. Don't remember to create you own `marked.Renderer` instance and make config right.

**Comment by joshbruce** [comment]
#952

**Comment by joshbruce** [comment]
Closing as it appears a fix was merged, but may not be working.
